### PR TITLE
feat: add rip support

### PR DESCRIPTION
**Pull request by markharden817**
# Description

Adding RIP configuration objects to be added to `panos.network.VirtualRouter` instance. 

The following classes have been added to enable this functionality:
- Rip
- RipInterface
- RipAuthProfile
- RipAuthProfileMd5
- RipExportRules

Added Unit tests to cover the additional classes
- TestRip
- TestRipAuthProfile
- TestRipAuthProfileMd5
- TestRipInterface
- TestRipExportRules

Feature enhancement  [#329](https://github.com/PaloAltoNetworks/pan-os-python/issues/329) 

# Motivation and Context

Current network design requires RIP routing configuration on VirtualRouter objects. 

# How Has This Been Tested?

pylint pass
flake8 pass 
pytest live tests pass

image

### Functionality tested with the following driver script:

```
import os

from panos.firewall import Firewall
from panos.network import (
    RedistributionProfile,
    Rip,
    RipAuthProfile,
    RipAuthProfileMd5,
    RipExportRules,
    RipInterface,
    VirtualRouter,
)

HOSTNAME = os.environ["PAN_HOSTNAME"]
USERNAME = os.environ["PAN_USERNAME"]
PASSWORD = os.environ["PAN_PASSWORD"]

VR_NAME = "vr_1"
REDIST_NAME = "redist_1"
VR_INTERFACES = ["ethernet1/1"]
REDIST_INTERFACE = "ethernet1/1"


fw = Firewall(HOSTNAME, USERNAME, PASSWORD)

# find or create a virtual router
vr = fw.find_or_create(VR_NAME, VirtualRouter, interface=VR_INTERFACES)

# create redist profile
redist_profile = RedistributionProfile(
    name=REDIST_NAME, priority=1, action="redist"
)
vr.add(redist_profile)

rip_spec = {
    "enable": True,
    "reject_default_route": True,
    "allow_redist_default_route": True,
    "delete_intervals": 121,
    "expire_intervals": 181,
    "interval_seconds": 2,
    "update_intervals": 31,
}
rip = Rip(**rip_spec)

# add rip auth (password)
rip_auth = RipAuthProfile(
    name="rip_profile_1", auth_type="password", password="#Password1"
)
rip.add(rip_auth)

# add rip auth (md5)
rip_auth = RipAuthProfile(name="rip_profile_2", auth_type="md5")
md5 = RipAuthProfileMd5(keyid=1, key="#Password1", preferred=True)
rip_auth.add(md5)
rip.add(rip_auth)

# add rip export rules
rip_export = RipExportRule(name=REDIST_NAME, metric=10)
rip.add(rip_export)

# add rip interfaces
rip_interface_spec = {
    "name": REDIST_INTERFACE,
    "enable": True,
    "advertise_default_route": "advertise",
    "metric": 11,
    "auth_profile": "rip_profile_1",
    "mode": "passive",
}
rip_interface = RipInterface(**rip_interface_spec)
rip.add(rip_interface)

# add rip config to virtual router and apply changes
vr.add(rip)
vr.apply()

```
### Result

image
image
image
image

# Types of changes

- New feature (non-breaking change which adds functionality)

# Checklist

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes if appropriate.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by markharden817** [comment]
Hello @btorresgil @shinmog 

Would you please review my PR and comment if any changes are required.

Thank you!

File: panos/network.py [hunk 1]
```diff
@@ -2125,6 +2126,190 @@ def _setup(self):
...
+        params.append(
+            VersionedParamPath(
+                "update_intervals", path="/timers/update-intervals", vartype="int"
+            )
+        )
+
+        self._params = tuple(params)
+
+
+class RipInterface(VersionedPanObject):
```
Comment (shinmog) [inline panos/network.py:2199]: This class is missing a lot of params.  There is also a `choice` here at path `default-route` which should have its own variable (see `objects.AddressObjects` for an example of how to handle choices, they need their own param).

File: panos/network.py [hunk 2]
```diff
@@ -2125,6 +2126,190 @@ def _setup(self):
...
+        self._xpaths.add_profile(value="/auth-profile")
+
+        params = []
+        params.append(VersionedParamPath("name"))
+        params.append(
+            VersionedParamPath("type", values=["password", "md5"], path="{type}")
+        )
+        params.append(
+            VersionedParamPath(
+                "password", condition={"type": "password"}, path="{type}"
```
Comment (shinmog) [inline panos/network.py:2261]: `vartype="encrypted"`

File: panos/network.py [hunk 3]
```diff
@@ -2125,6 +2126,190 @@ def _setup(self):
...
+
+        self._params = tuple(params)
+
+
+class RipAuthProfile(VersionedPanObject):
+    """Rip Authentication Profile
+
+    Args:
+        name (str): Name of Auth Profile
+        type (str): 'password' or 'md5'
```
Comment (shinmog) [inline panos/network.py:2242]: Please rename this param (e.g. - `auth_type`).  Naming variables `type` is trouble.

**Review by shinmog — CHANGES_REQUESTED** [review]
This is confusing and something that we'll need to enhance tests to check for, but when specifying the xpath of an object it needs to start with a `/`, however when specifying the xpath for a param it shouldn't.  I saw that this got you in a few places and have called them out.

**Comment by shinmog** [comment]
As an aside, I appreciate the addition of tests.

File: panos/network.py [hunk 4]
```diff
@@ -2125,6 +2126,190 @@ def _setup(self):
...
+            )
+        )
+        params.append(
+            VersionedParamPath(
+                "expire_intervals", path="/timers/expire-intervals", vartype="int"
+            )
+        )
+        params.append(
+            VersionedParamPath(
+                "interval_seconds", path="/timers/interval-seconds", vartype="int"
```
[the remark this answers is not in the document]
Comment (markharden817) [reply panos/network.py]: Removed leading param path `/`

File: panos/network.py [hunk 5]
```diff
@@ -2125,6 +2126,190 @@ def _setup(self):
...
+            )
+        )
+        params.append(
+            VersionedParamPath(
+                "delete_intervals", path="/timers/delete-intervals", vartype="int"
+            )
+        )
+        params.append(
+            VersionedParamPath(
+                "expire_intervals", path="/timers/expire-intervals", vartype="int"
```
[the remark this answers is not in the document]
Comment (markharden817) [reply panos/network.py]: Removed leading param path /

File: panos/network.py [hunk 6]
```diff
@@ -2125,6 +2126,190 @@ def _setup(self):
...
+        params.append(
+            VersionedParamPath(
+                "allow_redist_default_route",
+                path="allow-redist-default-route",
+                vartype="yesno",
+            )
+        )
+        params.append(
+            VersionedParamPath(
+                "delete_intervals", path="/timers/delete-intervals", vartype="int"
```
[the remark this answers is not in the document]
Comment (markharden817) [reply panos/network.py]: Removed leading param path /

File: panos/network.py [hunk 7]
```diff
@@ -2125,6 +2126,190 @@ def _setup(self):
...
+    """
+
+    SUFFIX = ENTRY
+
+    def _setup(self):
+        self._xpaths.add_profile(value="/interface")
+
+        params = []
+
+        params.append(VersionedParamPath("enable", path="enable", vartype="yesno"))
```
[the remark this answers is not in the document]
Comment (markharden817) [reply panos/network.py]: Added param default selection

File: panos/network.py (hunk 2 above)
Comment (markharden817) [reply panos/network.py]: added `vartype="encrypted"`

File: panos/network.py (hunk 3 above)
Comment (markharden817) [reply panos/network.py]: Renamed param

File: panos/network.py [hunk 8]
```diff
@@ -2125,6 +2126,190 @@ def _setup(self):
...
+            )
+        )
+        params.append(
+            VersionedParamPath(
+                "interval_seconds", path="/timers/interval-seconds", vartype="int"
+            )
+        )
+        params.append(
+            VersionedParamPath(
+                "update_intervals", path="/timers/update-intervals", vartype="int"
```
[the remark this answers is not in the document]
Comment (markharden817) [reply panos/network.py]: Added `default=30` and removed leading param path /

**Comment by markharden817** [comment]
> As an aside, I appreciate the addition of tests.

I really like your testing framework. It made easy to onboard additional test cases.

**Review by shinmog — APPROVED** [review]
lgtm

**Comment by shinmog** [comment]
@markharden817 

Looks like you need to run `make format` still.

**Comment by shinmog** [comment]
@markharden817 

Some formatting is still off - `make format` fixes the formatting.

**Comment by markharden817** [comment]
@shinmog 
I am not getting any output from `make format`. I did find more `black` and `isort` format changes changes.

**Comment by markharden817** [comment]
@shinmog 

Ok, I have resolved the issue with `make format`.  Having issues with poetry on Windows but it is working in WSL. 

```
WSL - mark: pan-os-python on feat/add-rip-support [3.8.10] ❯ poetry run make format
isort --recursive --atomic panos
black .
reformatted /mnt/c/Users/Mark/Documents/projects/pan-os-python/tests/live/test_network.py
reformatted /mnt/c/Users/Mark/Documents/projects/pan-os-python/panos/network.py
All done! ✨ 🍰 ✨
```

**Pull request merged**